cli/listener/src/main.rs

```diff
@@ -210,74 +215,417 @@ pub fn get_apy_for_period(
+                help: "Number of times we polled the exchange rate, since launch.",
+                type_: "counter",
+                metrics: vec![Metric::new(self.polls)],
+            },
+        )?;
+        write_metric(out, &MetricFamily {
+            name: "solido_pricedb_errors_total",
+            help: "Number of times we encountered an error while trying to get the exchange rate, since launch.",
+            type_: "counter",
+            metrics: vec![Metric::new(self.errors)]
+        })?;
+        if let Some(interval_price) = &self.solido_average_interval_price {
+            write_metric(
+                out,
+                &MetricFamily {
+                    name: "average_apy",
+                    help: "Average APY",
+                    type_: "counter",
+                    metrics: vec![Metric::new(interval_price.annual_percentage_rate())
+                        .with_label("solido", "APY".to_string())],
+                },
+            )?;
+        }
+
+        Ok(())
+    }
+}
+
+fn serve_request(request: Request, metrics_mutex: &MetricsMutex) -> Result<(), std::io::Error> {
+    // Take the current snapshot. This only holds the lock briefly, and does
+    // not prevent other threads from updating the snapshot while this request
+    // handler is running.
+    let metrics = metrics_mutex.lock().unwrap().clone();
+
+    // We don't even look at the request, for now we always serve the metrics.
+
+    let mut out: Vec<u8> = Vec::new();
+    let is_ok = metrics.write_prometheus(&mut out).is_ok();
+
+    if is_ok {
+        // text/plain with version=0.0.4 is what Prometheus expects as the content type,
+        // see also https://prometheus.io/docs/instrumenting/exposition_formats/.
+        // We add the charset so you can view the metrics in a browser too when it
+        // contains non-ascii bytes.
+        let content_type = Header::from_bytes(
+            &b"Content-Type"[..],
+            &b"text/plain; version=0.0.4; charset=UTF-8"[..],
+        )
+        .expect("Static header value, does not fail at runtime.");
+        request.respond(Response::from_data(out).with_header(content_type))
+    } else {
+        request.respond(Response::from_string("error").with_status_code(500))
+    }
+}
+
+/// Spawn threads that run the http server.
+fn start_http_server(opts: &Opts, metrics_mutex: Arc<MetricsMutex>) -> Vec<JoinHandle<()>> {
+    let server = match Server::http(opts.listen.clone()) {
+        Ok(server) => Arc::new(server),
+        Err(err) => {
+            eprintln!(
+                "Error: {}\nFailed to start http server on {}. Is the daemon already running?",
+                err, &opts.listen,
+            );
+            std::process::exit(1);
+        }
+    };
+
+    println!("Http server listening on {}", &opts.listen);
+
+    // Spawn a number of http handler threads, so we can handle requests in
+    // parallel. This server is only used to serve metrics, it can be super basic,
```

Comment:
This will not really be true for this daemon though. Still, 1 thread per core should be fine. Maybe 2 threads per core if we expect them to be IO-bound. Maybe eventually we need to make it configurable on the command line, but for now 1 thread per core would be fine. We shouldn’t lie in the comment though.